solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol

```diff
@@ -0,0 +1,1362 @@
+    */
+    function doConvert(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, address _trader, address payable _beneficiary)
+        internal
+        returns (uint256)
+    {
+        // update the recent average rate
+        updateRecentAverageRate();
+
+        uint256 sourceId = __reserveIds[_sourceToken];
+        uint256 targetId = __reserveIds[_targetToken];
+
+        (uint256 sourceBalance, uint256 targetBalance) = readReserveBalances(sourceId, targetId);
+        uint256 targetAmount = crossReserveTargetAmount(sourceBalance, targetBalance, _amount);
+
+        // get the target amount minus the conversion fee and the conversion fee
+        uint256 fee = calculateFee(targetAmount);
+        uint256 amount = targetAmount - fee;
+
+        // ensure that the trade gives something in return
+        require(amount != 0, "ERR_ZERO_TARGET_AMOUNT");
+
+        // ensure that the trade won't deplete the reserve balance
+        assert(amount < targetBalance);
+
+        // ensure that the input amount was already deposited
+        uint256 actualSourceBalance;
+        if (_sourceToken == ETH_RESERVE_ADDRESS) {
+            actualSourceBalance = address(this).balance;
+            require(msg.value == _amount, "ERR_ETH_AMOUNT_MISMATCH");
+        }
+        else {
+            actualSourceBalance = _sourceToken.balanceOf(address(this));
+            require(msg.value == 0 && actualSourceBalance.sub(sourceBalance) >= _amount, "ERR_INVALID_AMOUNT");
+        }
+
+        // sync the reserve balances
+        writeReserveBalances(sourceId, targetId, actualSourceBalance, targetBalance - amount);
+
+        // transfer funds to the beneficiary in the to reserve token
+        if (_targetToken == ETH_RESERVE_ADDRESS)
+            _beneficiary.transfer(amount);
+        else
+            safeTransfer(_targetToken, _beneficiary, amount);
+
+        // dispatch the conversion event
+        dispatchConversionEvent(_sourceToken, _targetToken, _trader, _amount, amount, fee);
+
+        // dispatch rate updates
+        dispatchTokenRateUpdateEvents(_sourceToken, _targetToken, actualSourceBalance, targetBalance - amount);
+
+        return amount;
+    }
+
+    /**
+      * @dev returns the recent average rate of 1 `_token` in the other reserve token units
+      *
+      * @param _token   token to get the rate for
+      * @return recent average rate between the reserves (numerator)
+      * @return recent average rate between the reserves (denominator)
+    */
+    function recentAverageRate(IERC20Token _token) external view returns (uint256, uint256) {
+        // get the recent average rate of reserve 0
+        uint256 rate = calcRecentAverageRate(averageRateInfo);
+        if (_token == __reserveTokens[0]) {
+            return ((rate >> 112) & MAX_UINT112, rate & MAX_UINT112);
+        }
+
+        return (rate & MAX_UINT112, (rate >> 112) & MAX_UINT112);
```

Comment:
maybe better to add a utility instead of repeating these operations in multiple places?

Reply:
As far as I remember, this is actually the only place where `112` (and this code-pattern in general) appears.

Reply:
also used when calculating the average rate

Reply:
let's continue this in a single discussion, it's a generic concept.

Reply:
OK